k9mail/src/main/java/com/fsck/k9/ui/ContactBadge.java

```diff
@@ -0,0 +1,236 @@
+            onContactUriChanged();
+        }
+    }
+
+    private void onContactUriChanged() {
+        setEnabled(isAssigned());
+    }
+
+    @Override
+    public void onClick(View v) {
+        // If contact has been assigned, mExtras should no longer be null, but do a null check
+        // anyway just in case assignContactFromPhone or Email was called with a null bundle or
+        // wasn't assigned previously.
+        final Bundle extras = (this.extras == null) ? new Bundle() : this.extras;
+        if (contactUri != null) {
+            QuickContact.showQuickContact(getContext(), ContactBadge.this, contactUri,
+                    QuickContact.MODE_LARGE, null);
+        } else if (contactEmail != null) {
+            extras.putString(EXTRA_URI_CONTENT, contactEmail);
+            queryHandler.startQuery(TOKEN_EMAIL_LOOKUP_AND_TRIGGER, extras,
+                    Uri.withAppendedPath(Email.CONTENT_LOOKUP_URI, Uri.encode(contactEmail)),
+                    EMAIL_LOOKUP_PROJECTION, null, null, null);
+        }
+    }
+
+    @Override
+    public void onInitializeAccessibilityEvent(AccessibilityEvent event) {
+        super.onInitializeAccessibilityEvent(event);
+        event.setClassName(ContactBadge.class.getName());
+    }
+
+    @Override
+    public void onInitializeAccessibilityNodeInfo(AccessibilityNodeInfo info) {
+        super.onInitializeAccessibilityNodeInfo(info);
+        info.setClassName(ContactBadge.class.getName());
+    }
+
+    private class QueryHandler extends AsyncQueryHandler {
+
+        QueryHandler(ContentResolver cr) {
+            super(cr);
+        }
+
+        @Override
+        protected void onQueryComplete(int token, Object cookie, Cursor cursor) {
+            Uri lookupUri = null;
+            Uri createUri = null;
+            boolean trigger = false;
+            Bundle extras = (cookie != null) ? (Bundle) cookie : new Bundle();
+            try {
+                switch (token) {
+                    case TOKEN_EMAIL_LOOKUP_AND_TRIGGER:
+                        trigger = true;
+                        createUri = Uri.fromParts("mailto",
+                                extras.getString(EXTRA_URI_CONTENT), null);
+
+                        //$FALL-THROUGH$
+                    case TOKEN_EMAIL_LOOKUP: {
+                        if (cursor != null && cursor.moveToFirst()) {
+                            long contactId = cursor.getLong(EMAIL_ID_COLUMN_INDEX);
+                            String lookupKey = cursor.getString(EMAIL_LOOKUP_STRING_COLUMN_INDEX);
+                            lookupUri = Contacts.getLookupUri(contactId, lookupKey);
+                        }
+                        break;
+                    }
+                }
+            } finally {
+                if (cursor != null) {
+                    cursor.close();
+                }
+            }
+
+            contactUri = lookupUri;
+            onContactUriChanged();
+
+            if (trigger && lookupUri != null) {
```

Comment:
could move this `if() else` block into a seperate function to keep this more less massive.